pkg/build/builder/common.go

```diff
@@ -82,3 +92,124 @@ func updateBuildRevision(c client.BuildInterface, build *api.Build, sourceInfo *
 		glog.Warningf("An error occurred saving build revision: %v", err)
 	}
 }
+
+// randomBuildTag generates a random tag used for building images in such a way
+// that the built image can be referred to unambiguously even in the face of
+// concurrent builds with the same name in the same namespace.
```

Comment:
As I said, `docker build -t repo:tag .` always builds and overwrites whatever is in `repo:tag`.

Remember that `docker build` and `docker tag` steps will run even if there is no post commit hook to be executed. We cannot rely on the hook failure because of container name collision.